Multiple improvements to inode_link

mykaul opened the pull request:
- [x] Improve inode_link(): pass table->name_hash[hash] list_head
    Instead of passing just the index, pass table->name_hash[index] (list_head structure) to the functions that use it.
    
- [x] Improve inode_link(): move the error logging allocation to where it's needed.
    No need to declare and certainly not memset, by default, two 64 byte arrays for an error message.
    Do it when needed.

- [x] Improve inode_link(): extract checks outside the lock, before the action
    There are multiple checks taking place under the lock on __inode_link().
    They were extracted to pre_inode_link_checks() function that takes place before.
    
- [x] Improve inode_link(): remove __dentry_hash()
    No need to list_del anything - it's empty already.
    
- [x] Improve inode_link(): Remove inode_hash()
    No need to list_del(inode->hash) - it's empty already, sincde __is_inode_hashed() returned false earlier.
    So just called directly the one liner left - adding it to the table inode_hash in the right location.

Updates: #3216
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
Improve inode_link(): improve dentry_create()

- Use mem_get() instead of mem_get0(), as we populate all fields.
- Populate all fields, in the order they appear in the structure.

Updates: #3216
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
Improve inode_link(): Remove inode_hash()

- no need to list_del(inode->hash) - it's empty already, sincde __is_inode_hashed() returned false earlier.
So just called directly the one liner left - adding it to the table inode_hash in the right location.

Updates: #3216
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
Improve inode_link(): remove __dentry_hash()
- no need to list_del anything - it's empty already.

Updates: #3216
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
Improve inode_link(): extract checks outside the lock, before the action

There are multiple checks taking place under the lock on __inode_link().
They were extracted to pre_inode_link_checks() function that takes place before.

Updates: #3216
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

Commit:
Improve inode_link(): move the error logging allocation to where it's needed.

No need to declare and certainly not memset, by default, two 64 byte arrays for an error message.
Do it when needed.

Updates: #3216
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
/run regression

amarts commented on rpc/rpc-lib/src/rpc-clnt.c:
```diff
@@ -40,25 +40,24 @@ __saved_frames_get_timedout(struct saved_frames *frames, time_t latest)
-    if (SFRAME_GET_PROGNUM(sframe) == GLUSTER_FOP_PROGRAM &&
-        SFRAME_GET_PROGVER(sframe) == GLUSTER_FOP_VERSION)
-        fop = SFRAME_GET_PROCNUM(sframe);
+    if (RPCREQ_GET_PROGNUM(rpcreq) == GLUSTER_FOP_PROGRAM &&
```
This change looks to be part of #3146 .... not related to inode link related.

amarts reviewed (commented):
inode.c changes looks interesting, and I feel fine to merge that. @xhernandez @pranithk have a look when you get a chance.

xhernandez reviewed (commented):
LGTM

Commit:
Improve inode_link(): pass table->name_hash[hash] list_head

Instead of passing just the index, pass table->name_hash[index] (list_head structure) to the functions that use it.

Updates: #3216
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
/run regression

gluster-ant commented:
0 test(s) failed 


1 test(s) generated core 
./tests/bugs/glusterd/brick-mux-validation-in-cluster.t

4 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/bugs/bitrot/bug-1245981.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
https://build.gluster.org/job/gh_centos7-regression/2220/

mykaul commented:
/run regression

Commit:
libglusterfs/src/inode.c: make __dentry_grep static

It's only used within this file.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

mykaul commented:
/run regression

Commit:
inode.c: remove dead code - inode_grep_for_gfid()

Not used by anyone.

Updates: #1000
Signed-off-by: Yaniv Kaul <ykaul@redhat.com>

jkroonza commented:
@mykaul I'm trying to apply this on top of v10.2 but it doesn't apply cleanly, and I suspect this relies on stuff which is not yet part of the release-10 branch.  Does the logic changes apply to release-10 as well?

I'd really prefer to not run a full devel branch on production, but it does look like there are lock changes involved in this PR which may well affect what I'm seeing in #3659 (or at least suspect).

mykaul commented:
> @mykaul I'm trying to apply this on top of v10.2 but it doesn't apply cleanly, and I suspect this relies on stuff which is not yet part of the release-10 branch. Does the logic changes apply to release-10 as well?

It's a bunch of small changes, but certainly not trivial (which is why it's not merged or reviewed yet). 
I doubt we'll backport any to 10.x, unless they prove very very useful - such as in your case.

> 
> I'd really prefer to not run a full devel branch on production, but it does look like there are lock changes involved in this PR which may well affect what I'm seeing in #3659 (or at least suspect).

Can you identify which you suspect are improving things? I have at least 2 PRs in the works, with more small changes.

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/inode.c b/libglusterfs/src/inode.c
index beb18dd1c..0e4914697 100644
--- a/libglusterfs/src/inode.c
+++ b/libglusterfs/src/inode.c
@@ -1637,7 +1637,7 @@ __inode_table_init_root(inode_table_t *table)
 {
     inode_t *root = NULL;
     static uuid_t root_gfid = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1};
-    
+
     root = inode_create(table);
 
     list_add(&root->list, &table->lru);
@@ -1701,7 +1701,8 @@ inode_table_with_invalidator(uint32_t lru_limit, xlator_t *xl,
         if (inode_hashsize > 0)
             gf_log(THIS->name, GF_LOG_WARNING,
                    "Set inode table size to minimum value of 65536 rather than "
-                   "the configured %u", inode_hashsize);
+                   "the configured %u",
+                   inode_hashsize);
         new->inode_hashsize = 65536;
     } else {
         new->inode_hashsize = inode_hashsize;
```

mykaul commented:
Need to properly fix merge conflicts :/

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/inode.c b/libglusterfs/src/inode.c
index 780eb6dde..8c45cf3cb 100644
--- a/libglusterfs/src/inode.c
+++ b/libglusterfs/src/inode.c
@@ -1667,7 +1667,8 @@ inode_table_with_invalidator(uint32_t lru_limit, xlator_t *xl,
         if (inode_hashsize > 0)
             gf_log(THIS->name, GF_LOG_WARNING,
                    "Set inode table size to minimum value of 65536 rather than "
-                   "the configured %u", inode_hashsize);
+                   "the configured %u",
+                   inode_hashsize);
         new->inode_hashsize = 65536;
     } else {
         new->inode_hashsize = inode_hashsize;
```

Commit:
Fix compilation

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/inode.c b/libglusterfs/src/inode.c
index 78a40f624..c96c64284 100644
--- a/libglusterfs/src/inode.c
+++ b/libglusterfs/src/inode.c
@@ -1612,7 +1612,8 @@ __inode_table_init_root(inode_table_t *table)
 
     gf_uuid_copy(root->gfid, root_gfid);
     root->ia_type = IA_IFDIR;
-    list_add(&root->hash, &table->inode_hash[hash_gfid(root_gfid, table->inode_hashsize)]);
+    list_add(&root->hash,
+             &table->inode_hash[hash_gfid(root_gfid, table->inode_hashsize)]);
 
     root = __inode_ref(root, _gf_false);
 
@@ -1667,7 +1668,8 @@ inode_table_with_invalidator(uint32_t lru_limit, xlator_t *xl,
         if (inode_hashsize > 0)
             gf_log(THIS->name, GF_LOG_WARNING,
                    "Set inode table size to minimum value of 65536 rather than "
-                   "the configured %u", inode_hashsize);
+                   "the configured %u",
+                   inode_hashsize);
         new->inode_hashsize = 65536;
     } else {
         new->inode_hashsize = inode_hashsize;
```

mykaul commented:
/run regression

Commit:
clang-format fixes

mykaul commented:
/run regression

mykaul commented:
This is now ready (again) for review, I hope I resolved conflicts properly.

mykaul committed:
Merge branch 'devel' into inode_dentry_create

mykaul commented:
> This is now ready (again) for review, I hope I resolved conflicts properly.

More conflicts resolved (which I think are from a different PR I've sent a while ago). I hope they were properly resolved!

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/glusterfs/inode.h b/libglusterfs/src/glusterfs/inode.h
index 935dfec57..ce0900304 100644
--- a/libglusterfs/src/glusterfs/inode.h
+++ b/libglusterfs/src/glusterfs/inode.h
@@ -73,7 +73,7 @@ struct _dentry {
     struct list_head hash;       /* hash table pointers */
     inode_t *inode;              /* inode of this directory entry */
     inode_t *parent;             /* directory of the entry */
-    char name[];                  /* name of the directory entry */
+    char name[];                 /* name of the directory entry */
 };
 
 struct _inode_ctx {
```

mykaul commented:
/run regression
